nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java

```diff
@@ -0,0 +1,239 @@
+            logger.info("Was disconnected from client or was never connected, attempting to connect.");
+            synchronized (mqttClientConnectLock){
+                try {
+                    reconnect();
+                } catch (MqttException e) {
+                    context.yield();
+                    session.transfer(flowfile, REL_FAILURE);
+                    logger.error("MQTT client is disconnected and re-connecting failed. Transferring FlowFile to fail and yielding", e);
+                    return;
+                }
+            }
+        }
+
+        // get the MQTT topic
+        String topic = context.getProperty(PROP_TOPIC).evaluateAttributeExpressions(flowfile).getValue();
+
+        if (topic == null || topic.isEmpty()) {
+            logger.warn("Evaluation of the topic property returned null or evaluated to be empty, routing to failure");
+            session.transfer(flowfile, REL_FAILURE);
+            return;
+        }
+
+        // do the read
+        final byte[] messageContent = new byte[(int) flowfile.getSize()];
+        session.read(flowfile, new InputStreamCallback() {
+            @Override
+            public void process(final InputStream in) throws IOException {
+                StreamUtils.fillBuffer(in, messageContent, true);
+            }
+        });
+
+        int qos = context.getProperty(PROP_QOS).evaluateAttributeExpressions(flowfile).asInteger();
+        final MqttMessage mqttMessage = new MqttMessage(messageContent);
+        mqttMessage.setQos(qos);
+        mqttMessage.setPayload(messageContent);
+        mqttMessage.setRetained(context.getProperty(PROP_RETAIN).evaluateAttributeExpressions(flowfile).asBoolean());
+
+        try {
+            /*
+             * Underlying method waits for the message to publish (according to set QoS), so it executes synchronously:
+             *     MqttClient.java:361 aClient.publish(topic, message, null, null).waitForCompletion(getTimeToWait());
+             */
+            mqttClient.publish(topic, mqttMessage);
+            session.transfer(flowfile, REL_SUCCESS);
+        } catch(MqttException me) {
+            logger.error("Failed to publish message.", me);
+            session.transfer(flowfile, REL_FAILURE);
+        }
+    }
+
+    private class PublishMQTTCallback  implements MqttCallback {
+
+        @Override
+        public void connectionLost(Throwable cause) {
+            logger.warn("Connection to " + broker + " lost", cause);
+            try {
+                reconnect();
+            } catch (MqttException e) {
+                logger.error("Connection to " + broker + " lost and re-connect failed");
+            }
+        }
+
+        @Override
+        public void messageArrived(String topic, MqttMessage message) throws Exception {
+            logger.error("Message arrived to a PublishMQTT processor { topic:'" + topic +"; payload:"+ Arrays.toString(message.getPayload())+"}");
```

Comment:
This is intentional because this is Publisher and really shouldn't ever receive a message. If it does then something weird is happening.